### ISSUE TITLE
Distribution pull through proxy 404 on all images

### ISSUE THREAD

**Issue by valentinbud** [docker/distribution#1448]
I have configured a pull through cache in my network to speed image downloads and offload Docker infra.

```
$ cat /docker-data/iproxy/config.yml
version: 0.1
log:
  level: debug
  formatter: text
  fields:
    service: registry
storage:
    cache:
        blobdescriptor: inmemory
    filesystem:
        rootdirectory: /var/lib/registry
http:
    addr: :5000
    headers:
        X-Content-Type-Options: [nosniff]
    tls:
        certificate: /var/lib/registry/__databus_systems.crt
        key: /var/lib/registry/STAR.databus.systems.key
health:
  storagedriver:
    enabled: true
    interval: 10s
    threshold: 3
proxy:
    remoteurl: https://registry-1.docker.io
    username: vbud
    password: ******
```

I only want to 'mirror' public images so first I have tried without `username` and `password` for `proxy` but after `curl https://registry-1.docker.io/v2/library/debian/manifests/sid` I thought it doesn't work because of that. Adding `username` and `password` doesn't change anything.

```
$ curl -I https://registry-1.docker.io/v2/library/debian/manifests/sid
HTTP/1.1 401 Unauthorized
Content-Type: application/json; charset=utf-8
Docker-Distribution-Api-Version: registry/2.0
Www-Authenticate: Bearer realm="https://auth.docker.io/token",service="registry.docker.io",scope="repository:library/debian:pull"
Date: Wed, 10 Feb 2016 08:30:17 GMT
Content-Length: 173
Strict-Transport-Security: max-age=31536000
```

Notice the `401 Unauthorized` response.

Also there is an NGINX proxy configured as a reverse proxy, running in a separate container with `--net=host` to simplify configuration mainly.

```
$ docker exec nginx-proxy cat /etc/nginx/conf.d/iproxy.databus.systems.conf
map $upstream_http_docker_distribution_api_version $docker_distribution_api_version {
    'registry/2.0' '';
    default registry/2.0;
}

server {
  listen 443 ssl;
  server_name iproxy.databus.systems;

  access_log /var/log/nginx/access.log;
  error_log /var/log/nginx/reg.error.log debug;

  # SSL
  ssl_certificate /private/STAR.databus.systems.crt;
  ssl_certificate_key /private/STAR.databus.systems.key;

  # Recommendations from https://raymii.org/s/tutorials/Strong_SSL_Security_On_nginx.html
  ssl_protocols TLSv1.1 TLSv1.2;
  ssl_ciphers 'EECDH+AESGCM:EDH+AESGCM:AES256+EECDH:AES256+EDH';
  ssl_prefer_server_ciphers on;
  ssl_session_cache shared:SSL:10m;

  # disable any limits to avoid HTTP 413 for large image uploads
  client_max_body_size 0;

  # required to avoid HTTP 411: see Issue #1486 (https://github.com/docker/docker/issues/1486)
  chunked_transfer_encoding on;

  location / {
      if ($http_user_agent ~ "^(docker\/1\.(3|4|5|10(?!\.[0-9]-dev))|Go ).*\$" ) {
        return 404;
      }

      # To add basic authentication to v2 use auth_basic setting plus add_header
      # auth_basic "Registry realm";
      # auth_basic_user_file    /private/iproxy.htpasswd;
      add_header 'Docker-Distribution-Api-Version' 'registry/2.0';

      proxy_pass                          http://iproxy;
      proxy_http_version 1.1;
      proxy_set_header  Connection "";
      proxy_set_header  Host              $http_host;   # required for docker client's sake
      proxy_set_header  X-Real-IP         $remote_addr; # pass on real client's IP
      proxy_set_header  X-Forwarded-For   $proxy_add_x_forwarded_for;
      proxy_set_header  X-Forwarded-Proto $scheme;
      proxy_read_timeout                  900;
  }
}
```

The NGINX configuration is the same as for a private registry without the authentication part. I have also added `proxy_http_version 1.1;` and `proxy_set_header  Connection "";`. The behaviour is the same with them or without them.

```
$ docker rmi -f debian:sid
$ docker pull debian:sid
$ docker logs iproxy
time="2016-02-10T08:35:53Z" level=warning msg="No HTTP secret provided - generated random secret. This may cause problems with uploads if multiple registries are behind a load-balancer. To provide a shared secret, fill in http.secret in the configuration file or set the REGISTRY_HTTP_SECRET environment variable." go.version=go1.5.3 instance.id=5ae36eb5-d22d-4605-8826-81e77489551b version=v2.3.0
time="2016-02-10T08:35:53Z" level=info msg="redis not configured" go.version=go1.5.3 instance.id=5ae36eb5-d22d-4605-8826-81e77489551b version=v2.3.0
time="2016-02-10T08:35:53Z" level=info msg="Starting upload purge in 1m0s" go.version=go1.5.3 instance.id=5ae36eb5-d22d-4605-8826-81e77489551b version=v2.3.0
time="2016-02-10T08:35:53Z" level=info msg="using inmemory blob descriptor cache" go.version=go1.5.3 instance.id=5ae36eb5-d22d-4605-8826-81e77489551b version=v2.3.0
time="2016-02-10T08:35:53Z" level=info msg="listening on [::]:5000" go.version=go1.5.3 instance.id=5ae36eb5-d22d-4605-8826-81e77489551b version=v2.3.0
time="2016-02-10T08:36:02Z" level=info msg="response completed" go.version=go1.5.3 http.request.host=iproxy.databus.systems http.request.id=e69f31f5-3610-4e90-9727-64533cbcc1ad http.request.method=GET http.request.remoteaddr=86.106.37.0 http.request.uri="/v2/" http.request.useragent="docker/1.10.0 go/go1.5.3 git-commit/590d5108 kernel/4.1.17-boot2docker os/linux arch/amd64" http.response.contenttype="application/json; charset=utf-8" http.response.duration=4.568659ms http.response.status=200 http.response.written=2 instance.id=5ae36eb5-d22d-4605-8826-81e77489551b version=v2.3.0
127.0.0.1 - - [10/Feb/2016:08:36:02 +0000] "GET /v2/ HTTP/1.1" 200 2 "" "docker/1.10.0 go/go1.5.3 git-commit/590d5108 kernel/4.1.17-boot2docker os/linux arch/amd64"
time="2016-02-10T08:36:02Z" level=error msg="response completed with error" err.code="manifest unknown" err.detail="unknown tag=sid" err.message="manifest unknown" go.version=go1.5.3 http.request.host=iproxy.databus.systems http.request.id=117b8a92-3e79-4c9b-a99e-29012c26ba93 http.request.method=GET http.request.remoteaddr=86.106.37.0 http.request.uri="/v2/library/debian/manifests/sid" http.request.useragent="docker/1.10.0 go/go1.5.3 git-commit/590d5108 kernel/4.1.17-boot2docker os/linux arch/amd64" http.response.contenttype="application/json; charset=utf-8" http.response.duration=8.42156ms http.response.status=404 http.response.written=93 instance.id=5ae36eb5-d22d-4605-8826-81e77489551b vars.name="library/debian" vars.reference=sid version=v2.3.0
127.0.0.1 - - [10/Feb/2016:08:36:02 +0000] "GET /v2/library/debian/manifests/sid HTTP/1.1" 404 93 "" "docker/1.10.0 go/go1.5.3 git-commit/590d5108 kernel/4.1.17-boot2docker os/linux arch/amd64"
```

What's wrong with the above configuration? Thanks.

**Comment by RichardScothern** [comment]
Have configured the daemon to point to your registry?
https://docs.docker.com/registry/mirror/

**Comment by valentinbud** [comment]
The first client is a local VirtualBox Boot2Docker VM created with `docker-machine`.

```
$ docker-machine create --engine-registry-mirror https://iproxy.databus.systems --driver virtualbox dev
```

So if I knock with `curl` I have to authenticate and if I do it with Go client from `registry:2` I don't have to? I composed the `curl` URL by concatenating `proxy:remoteurl` with what I have found in `iproxy` logs.

```
127.0.0.1 - - [10/Feb/2016:08:36:02 +0000] "GET **/v2/library/debian/manifests/sid**
```

The second client is the host `registry:2` is hosted on. That's Debian Jessie. According to [Control and configure Docker with systemd](https://docs.docker.com/engine/admin/systemd/) I have the following configuration.

```
$ cat /etc/systemd/system/docker.service.d/registry-mirror.conf
[Service]
EnvironmentFile=-/etc/default/docker
ExecStart=
ExecStart=/usr/bin/docker daemon -H fd:// $DOCKER_OPTS

$ cat /etc/default/docker 
DOCKER_OPTS="--registry-mirror=https://iproxy.databus.systems"
```

I honestly tell I didn't do it anything but it stared working just a few seconds ago. Thank very much for your patience and assistance.

It works both with and without `username` and `password`.
